p2p/host/basic/basic_host_test.go

```diff
@@ -777,6 +778,49 @@ func TestHostAddrChangeDetection(t *testing.T) {
 	}
 }
 
+func TestNegotiationCancel(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	h1, h2 := getHostPair(ctx, t)
+	defer h1.Close()
+	defer h2.Close()
+
+	// pre-negotiation so we can make the negotiation hang.
+	h1.Network().SetStreamHandler(func(s network.Stream) {
+		<-ctx.Done() // wait till the test is done.
+		s.Reset()
+	})
+
+	ctx2, cancel2 := context.WithCancel(ctx)
+	defer cancel2()
+
+	errCh := make(chan error, 1)
+	go func() {
+		s, err := h2.NewStream(ctx2, h1.ID(), "/testing")
+		if s != nil {
+			errCh <- fmt.Errorf("expected to fail negotiation")
+			return
+		}
+		errCh <- err
+	}()
+	select {
+	case err := <-errCh:
+		t.Fatal(err)
+	case <-time.After(10 * time.Millisecond):
+		// ok, hung.
+	}
+	cancel2()
+
+	select {
+	case err := <-errCh:
+		require.Equal(t, err, context.Canceled)
+	case <-time.After(10 * time.Millisecond):
```

Comment:
Can we increase the timeout to 500ms to be on the safe side? The cancellation happens asynchronously, and 10ms might be a bit tight in some cases.

Reply:
10ms is a _long_ time for local operations. I'd err on the side of fixing this later if it's an issue, and keeping the tests fast for now. (but I don't feel strongly either way)

Reply:
SGTM, but flaky tests cause frustration. I'd rather have an (almost) imperceptible lag and never have to worry about it, but fine to keep it as is too and going your way. It's really a nit.

Reply:
Ok, I'll split the difference.

Reply:
(100ms)

Reply:
Oh, wait. Nvm. 500ms is the _timeout_ (it won't take 500ms in the happy path). Changed.